### DISCUSSION TITLE
hi all. unable to use opensc-pkcs11.so as a PHP library

### DISCUSSION THREAD

**Discussion by tamandua-rj** [OpenSC/OpenSC#2946] [General]
when loading PHP 5.6 I get this message:
PHP Warning:  PHP Startup: Invalid library (maybe not a PHP library) '/usr/lib/x86_64-linux-gnu/opensc-pkcs11.so'

opensc-tool  is ok and opensc-tool -a outputs this:
OpenSC 0.23.0 [gcc  13.2.0]
Enabled features: locking zlib readline openssl pcsc(libpcsclite.so.1)

My browser (firefox) has it's pkcs11 as it's pkcs11 (device manager) end recognizes my smart card correctly

any help will be appreciated

**Comment by Jakuje** [comment]
What are you trying to achieve? How do you load opensc into php? PHP5 is EOL for ages anyway:

https://www.php.net/supported-versions.php

**Comment by popovec** [comment]
The opensc pkcs#11 module will definitely not be loaded directly into php.. but it seems that there is a way:

https://github.com/gamringer/php-pkcs11

**Reply by tamandua-rj** [reply]
tks, wiil try it today

**Reply by tamandua-rj** [reply]
took a look and it is for  HSM only

**Comment by popovec** [comment]
Try changing the path to softhsm to the path to the opensc library.
`````
$modulePath = '/path/to/libsofthsm2.so';
`````
example php file:
`````
<?php
$modulePath = '/usr/lib/x86_64-linux-gnu/opensc-pkcs11.so';
$module = new Pkcs11\Module($modulePath);
$moduleInfo = $module->getInfo();
print_r ($moduleInfo);
?>
`````

then run:  `php -dextension=./modules/pkcs11.so -f test.php`

`````
Array
(
    [cryptokiVersion] => Array
        (
            [major] => 2
            [minor] => 20
        )

    [manufacturerID] => OpenSC Project
    [libraryDescription] => OpenSC smartcard framework
    [libraryVersion] => Array
        (
            [major] => 0
            [minor] => 24
        )

)
`````

**Comment by tamandua-rj** [comment]
![Screenshot from 2023-11-29 15-21-58](https://github.com/OpenSC/OpenSC/assets/33358048/aae30038-7625-47b1-8628-a9e479b52649)

**Comment by tamandua-rj** [comment]
![Screenshot from 2023-11-29 15-23-05](https://github.com/OpenSC/OpenSC/assets/33358048/627563a8-64f0-486e-92f9-9c565314218b)

**Reply by tamandua-rj** [reply]
did not work. php -dextension= should not point to where pkcs11.so is?? should it point to /usr/lib/x86_64-linux-gnu/opensc-pkcs11.so ??
thanks

**Comment by popovec** [comment]
First you have to compile the pkcs11 module for php, here is a complete procedure for linux:
`````
mkdir test
cd test
git clone https://github.com/gamringer/php-pkcs11
cd php-pkcs11
phpize
./configure
make
`````
if everything went correctly, your newly compiled module will be available:
`````
$ ls  modules/pkcs11.so 
modules/pkcs11.so
`````
Now you can run the test I mentioned above.

**Reply by tamandua-rj** [reply]
trying this but compiler is trowing a error.. will speak with gamringer people

**Comment by tamandua-rj** [comment]
Hi, can opensc be used with javascript?? have a example?? thanks

**Reply by popovec** [reply]
https://github.com/PeculiarVentures/pkcs11js
